File: scraper/app/model_catalog.py
"""Preference-ordered text models for Gemini / OpenAI / Anthropic.

Live discovery filters this list to IDs the project key can actually call.
"""
from __future__ import annotations

from typing import Iterable, Literal

CatalogProvider = Literal["gemini", "openai", "anthropic"]

DEFAULT_TEXT_MODEL = "gemini-2.5-flash"

GEMINI_TEXT_MODELS: tuple[str, ...] = (
    "gemini-flash-latest",
    "gemini-2.5-flash",
    "gemini-2.5-flash-lite",
    "gemini-2.0-flash",
    "gemini-2.0-flash-lite",
    "gemini-3.5-flash",
    "gemini-3.6-flash",
    "gemini-3.7-flash",
    "gemini-3.5-flash-lite",
    "gemini-3.1-flash-lite",
    "gemini-2.5-pro",
    "gemini-pro-latest",
    "gemini-3.1-pro-preview",
    "gemini-1.5-flash",
    "gemini-1.5-pro",
)

OPENAI_TEXT_MODELS: tuple[str, ...] = (
    "gpt-4o-mini",
    "gpt-4.1-mini",
    "gpt-4o",
    "gpt-4.1",
    "gpt-4-turbo",
)

ANTHROPIC_TEXT_MODELS: tuple[str, ...] = (
    "claude-3-haiku-20240307",
    "claude-3-5-haiku-20241022",
    "claude-3-5-sonnet-20241022",
    "claude-sonnet-4-20250514",
    "claude-3-7-sonnet-20250219",
)

APP_TO_API: dict[str, str] = {
    "gemini-flash": "gemini-2.5-flash",
    "gemini-pro": "gemini-2.5-pro",
    "claude": "claude-3-5-sonnet-20241022",
    "claude-3-5-sonnet": "claude-3-5-sonnet-20241022",
    "claude-3-haiku": "claude-3-haiku-20240307",
}

MAX_MODELS_PER_PROVIDER = 3
MAX_FALLBACK_MODELS = 9

_NON_TEXT = (
    "embed",
    "imagen",
    "tts",
    "whisper",
    "audio",
    "realtime",
    "transcribe",
    "moderation",
    "dall-e",
    "veo",
    "computer-use",
)


def strip_model_prefix(model_id: str) -> str:
    return model_id.strip().removeprefix("models/")


def provider_for_model(model: str) -> CatalogProvider | None:
    ident = strip_model_prefix(model).lower()
    if ident.startswith(("gpt-", "o1", "o3", "o4", "chatgpt")):
        return "openai"
    if ident.startswith("claude"):
        return "anthropic"
    if ident.startswith("gemini"):
        return "gemini"
    return None


def is_text_generation_model(model: str) -> bool:
    ident = strip_model_prefix(model).lower()
    if not ident:
        return False
    if any(token in ident for token in _NON_TEXT):
        return False
    if "image" in ident:
        return False
    return provider_for_model(ident) is not None


def rank_text_model(model: str) -> int:
    ident = strip_model_prefix(model)
    catalogs: dict[CatalogProvider, tuple[str, ...]] = {
        "gemini": GEMINI_TEXT_MODELS,
        "openai": OPENAI_TEXT_MODELS,
        "anthropic": ANTHROPIC_TEXT_MODELS,
    }
    provider = provider_for_model(ident)
    if provider:
        try:
            return catalogs[provider].index(ident)
        except ValueError:
            pass
    lowered = ident.lower()
    if "preview" in lowered or "-exp" in lowered or "experimental" in lowered:
        return 200
    if "lite" in lowered or "mini" in lowered or "haiku" in lowered:
        return 40
    if "flash" in lowered:
        return 50
    if "pro" in lowered or "sonnet" in lowered or "gpt-4" in lowered:
        return 80
    return 120


def merge_available(
    preferred: Iterable[str],
    available: set[str] | frozenset[str] | None,
) -> list[str]:
    preferred_list = list(preferred)
    if not available:
        return preferred_list
    out: list[str] = []
    seen: set[str] = set()
    for ident in preferred_list:
        if ident in available and ident not in seen and is_text_generation_model(ident):
            out.append(ident)
            seen.add(ident)
    extras = sorted(
        (
            ident
            for ident in available
            if ident not in seen and is_text_generation_model(ident)
        ),
        key=lambda ident: (rank_text_model(ident), ident),
    )
    for ident in extras:
        out.append(ident)
        seen.add(ident)
    return out


def build_fallback_chain(
    primary: str,
    *,
    gemini: bool = False,
    openai: bool = False,
    anthropic: bool = False,
    available_gemini: set[str] | None = None,
    available_openai: set[str] | None = None,
    available_anthropic: set[str] | None = None,
) -> list[str]:
    mapped = strip_model_prefix(APP_TO_API.get(primary, primary) if primary else DEFAULT_TEXT_MODEL) or DEFAULT_TEXT_MODEL
    primary_provider = provider_for_model(mapped)
    enabled: list[CatalogProvider] = []
    if gemini:
        enabled.append("gemini")
    if openai:
        enabled.append("openai")
    if anthropic:
        enabled.append("anthropic")

    order: list[CatalogProvider] = []
    if primary_provider and primary_provider in enabled:
        order.append(primary_provider)
    for provider in ("gemini", "openai", "anthropic"):
        if provider not in order and provider in enabled:
            order.append(provider)

    catalogs: dict[CatalogProvider, tuple[str, ...]] = {
        "gemini": GEMINI_TEXT_MODELS,
        "openai": OPENAI_TEXT_MODELS,
        "anthropic": ANTHROPIC_TEXT_MODELS,
    }
    available_map: dict[CatalogProvider, set[str] | None] = {
        "gemini": available_gemini,
        "openai": available_openai,
        "anthropic": available_anthropic,
    }
    keys = {"gemini": gemini, "openai": openai, "anthropic": anthropic}

    chain: list[str] = []
    seen: set[str] = set()

    def push(ident: str) -> None:
        if not ident or ident in seen or not is_text_generation_model(ident):
            return
        provider = provider_for_model(ident)
        if provider is None or not keys[provider]:
            return
        seen.add(ident)
        chain.append(ident)

    if primary_provider and keys.get(primary_provider):
        push(mapped)

    for provider in order:
        rest = merge_available(catalogs[provider], available_map[provider])[
            :MAX_MODELS_PER_PROVIDER
        ]
        for ident in rest:
            if len(chain) >= MAX_FALLBACK_MODELS:
                break
            push(ident)
        if len(chain) >= MAX_FALLBACK_MODELS:
            break

    if not chain and gemini:
        push(DEFAULT_TEXT_MODEL)
    return chain

File: supabase/functions/_shared/modelCatalog.ts
// Preference-ordered text models for Gemini / OpenAI / Anthropic.
// Live discovery filters this list to IDs the project key can actually call.
// Do not invent undocumented IDs — unknown live IDs are ranked, not hardcoded.

export type CatalogProvider = "gemini" | "openai" | "anthropic";

export const DEFAULT_TEXT_MODEL = "gemini-2.5-flash";

/** Cheap/fast first, then paid/pro. Preview/exp stay last. */
export const GEMINI_TEXT_MODELS: readonly string[] = [
  "gemini-flash-latest",
  "gemini-2.5-flash",
  "gemini-2.5-flash-lite",
  "gemini-2.0-flash",
  "gemini-2.0-flash-lite",
  "gemini-3.5-flash",
  "gemini-3.6-flash",
  "gemini-3.7-flash",
  "gemini-3.5-flash-lite",
  "gemini-3.1-flash-lite",
  "gemini-2.5-pro",
  "gemini-pro-latest",
  "gemini-3.1-pro-preview",
  "gemini-1.5-flash",
  "gemini-1.5-pro",
];

export const OPENAI_TEXT_MODELS: readonly string[] = [
  "gpt-4o-mini",
  "gpt-4.1-mini",
  "gpt-4o",
  "gpt-4.1",
  "gpt-4-turbo",
];

export const ANTHROPIC_TEXT_MODELS: readonly string[] = [
  "claude-3-haiku-20240307",
  "claude-3-5-haiku-20241022",
  "claude-3-5-sonnet-20241022",
  "claude-sonnet-4-20250514",
  "claude-3-7-sonnet-20250219",
];

/** App / profile slug → API model id */
export const APP_TO_API: Record<string, string> = {
  "gemini-flash": "gemini-2.5-flash",
  "gemini-pro": "gemini-2.5-pro",
  "gemini-2.0-flash": "gemini-2.0-flash",
  "gemini-2.0-flash-lite": "gemini-2.0-flash-lite",
  "gemini-2.5-flash": "gemini-2.5-flash",
  "gemini-2.5-flash-lite": "gemini-2.5-flash-lite",
  "gemini-2.5-pro": "gemini-2.5-pro",
  "gemini-flash-latest": "gemini-flash-latest",
  "gemini-1.5-pro": "gemini-2.5-pro",
  "gemini-1.5-flash": "gemini-2.5-flash",
  "gpt-4o": "gpt-4o",
  "gpt-4o-mini": "gpt-4o-mini",
  "gpt-4.1": "gpt-4.1",
  "gpt-4.1-mini": "gpt-4.1-mini",
  "gpt-4-turbo": "gpt-4-turbo",
  claude: "claude-3-5-sonnet-20241022",
  "claude-3-5-sonnet": "claude-3-5-sonnet-20241022",
  "claude-3-haiku": "claude-3-haiku-20240307",
  "claude-3-5-haiku": "claude-3-5-haiku-20241022",
};

export const MAX_MODELS_PER_PROVIDER = 3;
export const MAX_FALLBACK_MODELS = 9;

export function stripModelPrefix(id: string): string {
  return id.trim().replace(/^models\//, "");
}

export function providerForModel(model: string): CatalogProvider | null {
  const id = stripModelPrefix(model).toLowerCase();
  if (
    id.startsWith("gpt-") ||
    id.startsWith("o1") ||
    id.startsWith("o3") ||
    id.startsWith("o4") ||
    id.startsWith("chatgpt")
  ) {
    return "openai";
  }
  if (id.startsWith("claude")) return "anthropic";
  if (id.startsWith("gemini")) return "gemini";
  return null;
}

export function isTextGenerationModel(model: string): boolean {
  const n = stripModelPrefix(model).toLowerCase();
  if (!n) return false;
  if (
    /(embed|imagen|^image-|image-preview|-image$|-image-|tts|whisper|audio|realtime|transcribe|moderation|dall-e|veo|computer-use)/i
      .test(n)
  ) {
    return false;
  }
  return providerForModel(n) !== null;
}

export function rankTextModel(model: string): number {
  const id = stripModelPrefix(model);
  const catalogs: Record<CatalogProvider, readonly string[]> = {
    gemini: GEMINI_TEXT_MODELS,
    openai: OPENAI_TEXT_MODELS,
    anthropic: ANTHROPIC_TEXT_MODELS,
  };
  const provider = providerForModel(id);
  if (provider) {
    const idx = catalogs[provider].indexOf(id);
    if (idx >= 0) return idx;
  }
  const n = id.toLowerCase();
  if (n.includes("preview") || n.includes("-exp") || n.includes("experimental")) {
    return 200;
  }
  if (n.includes("lite") || n.includes("mini") || n.includes("haiku")) return 40;
  if (n.includes("flash")) return 50;
  if (n.includes("pro") || n.includes("sonnet") || n.includes("gpt-4")) return 80;
  return 120;
}

export function mapAppModelToApi(input: string): string {
  const trimmed = input.trim();
  if (!trimmed) return DEFAULT_TEXT_MODEL;
  return APP_TO_API[trimmed] ?? stripModelPrefix(trimmed);
}

export function mergeAvailable(
  preferred: readonly string[],
  available: ReadonlySet<string> | null | undefined,
): string[] {
  if (!available || available.size === 0) return [...preferred];
  const out: string[] = [];
  const seen = new Set<string>();
  for (const id of preferred) {
    if (available.has(id) && !seen.has(id) && isTextGenerationModel(id)) {
      out.push(id);
      seen.add(id);
    }
  }
  const extras = [...available]
    .filter((id) => !seen.has(id) && isTextGenerationModel(id))
    .sort((a, b) => rankTextModel(a) - rankTextModel(b) || a.localeCompare(b));
  for (const id of extras) {
    out.push(id);
    seen.add(id);
  }
  return out;
}

export interface FallbackKeys {
  gemini?: boolean;
  openai?: boolean;
  anthropic?: boolean;
}

export interface AvailableByProvider {
  gemini?: ReadonlySet<string> | null;
  openai?: ReadonlySet<string> | null;
  anthropic?: ReadonlySet<string> | null;
}

function take(models: string[], limit: number): string[] {
  return models.slice(0, Math.max(0, limit));
}

function providerModels(
  provider: CatalogProvider,
  available: AvailableByProvider | undefined,
): string[] {
  const preferred =
    provider === "gemini"
      ? GEMINI_TEXT_MODELS
      : provider === "openai"
        ? OPENAI_TEXT_MODELS
        : ANTHROPIC_TEXT_MODELS;
  return mergeAvailable(preferred, available?.[provider]);
}

/**
 * Primary first, then other models of that provider, then remaining providers
 * (cheap → paid). Filtered to live-available IDs when the catalog was listed.
 */
export function buildFallbackChain(
  primary: string,
  keys: FallbackKeys,
  available?: AvailableByProvider,
): string[] {
  const mapped = mapAppModelToApi(primary || DEFAULT_TEXT_MODEL);
  const primaryProvider = providerForModel(mapped);
  const enabled: CatalogProvider[] = [];
  if (keys.gemini) enabled.push("gemini");
  if (keys.openai) enabled.push("openai");
  if (keys.anthropic) enabled.push("anthropic");

  const order: CatalogProvider[] = [];
  if (primaryProvider && enabled.includes(primaryProvider)) {
    order.push(primaryProvider);
  }
  for (const p of ["gemini", "openai", "anthropic"] as const) {
    if (!order.includes(p) && enabled.includes(p)) order.push(p);
  }

  const chain: string[] = [];
  const seen = new Set<string>();
  const push = (id: string) => {
    if (!id || seen.has(id) || !isTextGenerationModel(id)) return;
    const provider = providerForModel(id);
    if (!provider || !keys[provider]) return;
    seen.add(id);
    chain.push(id);
  };

  if (primaryProvider && keys[primaryProvider]) push(mapped);

  for (const provider of order) {
    const rest = take(providerModels(provider, available), MAX_MODELS_PER_PROVIDER);
    for (const id of rest) {
      if (chain.length >= MAX_FALLBACK_MODELS) break;
      push(id);
    }
    if (chain.length >= MAX_FALLBACK_MODELS) break;
  }

  if (chain.length === 0 && keys.gemini) push(DEFAULT_TEXT_MODEL);
  return chain;
}
